### DISCUSSION TITLE
Profile-Guided optimizations and using other compilers?

### DISCUSSION THREAD

**Discussion by larsbohn** [ispc/ispc#2188] [Q&A] [answered]
Hi!

This might be a dumb question since I do not know how the optimization stages work with LLVM and ISPC.

Are there any ways to get profiling information into the ispc compile step? So in the lines of PGO used by the Intel C++ compiler?

And what if I use the Intel C++ Compiler generally for the project using the obj-files from ISPC - is any further performance improvement to be expected from this or will the Intel Compiler just pick up the obj-file as it is?

Thanks for any ideas...

**Comment by dbabokin** [comment, marked as answer]
ISPC doesn't support PGO.

With profile guided optimizations, the compiler produces the code that dumps execution counters on the first pass. Then user executes the code and get the profile. Finally, on the second pass, the compiler uses the profile to improve optimization decisions.

ISPC supports neither emitting nor consuming profiling informtion.

What would happen if only a part of object files in the project support PGO? Nothing criminal, that's actually pretty normal situations, as anything linked to the project in binary form would miss it (libc functions, for example).

What would be the impact on the performance, if ISPC supports PGO? While it could potentially improve performance, the effect would be way less than generally expected compared to C++. The most important PGO clients are
1. inter-procedural optimizer (i.e. inlining)
2. high-level optimizer (loop optimizations)
3. auto-vectorizer (directly or indirectly increasing aggressiveness in the hot loops)
4. it may also affect code scheduling (prioritizing some branches over the others)

(2) and (3) are not relevant for ISPC, as we rely on the user to express vector semantics explicitly and provide the idioms for loop parallelizations (`forech` loops), hence opportunities for high level optimizations are very limited.
(1) is typically less important, as ISPC represents computational kernels (as opposed to general purpose computations with hard-to-isolate computational kernels), which a-priori require more aggressive inlining.
And (4) is not that important on modern CPUs with out-of-order architectures.
